File: 205/isomorphic-strings.java
class Solution {
    char []tableST=new char[256];
    char []tableTS=new char[256];
    public boolean isIsomorphic(String s, String t) {
        int N = s.length();
        if(t.length()!=N) return false;

        for(int i=0; i<N; i++){
            char ss = s.charAt(i);
            char tt = t.charAt(i);
            if(tableST[ss]==0 && tableTS[tt]==0){
                tableST[ss]=tt;
                tableTS[tt]=ss;
            }else if(tableST[ss]==tt && tableTS[tt]==ss){
                continue;
            }else{
                return false;
            }
        }
        return true;
    }
}
//Case 4: s = "badc"
//t = "baba"

File: 205/isomorphic-strings.py
# 同構體，也就是「逐字結構相同」只是換字母而已
# (1) 字母出現頻率（分布狀況）相同，且(2) 一對一對應
# 註：稍早分析錯誤, 「誤以為」是字母出現的頻率(分布狀況)是否相同
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        N1, N2 = len(s), len(t)
        if N1 != N2: return False # 長度不同，一定錯

        table1 = collections.defaultdict(str) # s 對應到 t 的對照表
        table2 = collections.defaultdict(str) # t 對應到 s 的對照表
        for i in range(N1):
            if s[i] not in table1 and t[i] not in table2:
                table1[s[i]]=t[i] # 沒對照過，可建立對照表
                table2[t[i]]=s[i]
            elif table1[s[i]] != t[i] or table2[t[i]] != s[i]:
                return False # 本次對應 與之前不同，就錯了
        return True
# case 6/45: "badc" vs. "baba"
